packages/wrangler/src/core/define-command.ts

```diff
@@ -0,0 +1,163 @@
+	>
+>;
+
+export type HandlerContext = {
+	/**
+	 * The wrangler config file read from disk and parsed.
+	 * If no config file can be found, this value will undefined.
+	 * Set `behaviour.requireConfig` to refine this type and
+	 * throw if it cannot be found.
+	 */
+	config: Config;
+	/**
+	 * The logger instance provided to the command implementor as a convenience.
+	 */
+	logger: Logger;
+	/**
+	 * Error classes provided to the command implementor as a convenience
+	 * to aid discoverability and to encourage their usage.
+	 */
+	errors: {
+		UserError: typeof UserError;
+		FatalError: typeof FatalError;
+
+		// TODO: extend with other categories of error
+	};
+
+	// TODO: experiments
+	// TODO(future): API SDK
+
+	// TODO: prompt (cli package)
+	// TODO: accountId
+};
+
+export type CommandDefinition<
+	NamedArgs extends BaseNamedArgDefinitions = BaseNamedArgDefinitions,
+> = {
+	/**
+	 * The full command as it would be written by the user.
+	 */
+	command: Command;
+
+	/**
+	 * Descriptive information about the command which does not affect behaviour.
+	 * This is used for the CLI --help and subcommand --help output.
+	 * This should be used as the source-of-truth for status and ownership.
+	 */
+	metadata: Metadata;
+	/**
+	 * Controls shared behaviour across all commands.
+	 * This will allow wrangler commands to remain consistent and only diverge intentionally.
+	 */
+	behaviour?: {
+		/**
+		 * By default, metrics are sent if the user has opted-in.
+		 * Set this value to `false` to disable metrics unconditionally.
+		 *
+		 * @default true
+		 */
+		sendMetrics?: boolean;
+
+		/**
+		 * By default, wrangler's version banner will be printed before the handler is executed.
+		 * Set this value to `false` to skip printing the banner.
+		 *
+		 * @default true
+		 */
+		printBanner?: boolean;
+	};
+
+	/**
+	 * A plain key-value object describing the CLI args for this command.
+	 * Shared args can be defined as another plain object and spread into this.
+	 */
+	args: NamedArgs;
+	/**
+	 * Optionally declare some of the named args as positional args.
+	 * The order of this array is the order they are expected in the command.
+	 * Use args[key].demandOption and args[key].array to declare required and variadic
+	 * positional args, respectively.
+	 */
+	positionalArgs?: Array<StringKeyOf<NamedArgs>>;
+
+	/**
+	 * A hook to implement custom validation of the args before the handler is called.
+	 * Throw `CommandLineArgsError` with actionable error message if args are invalid.
+	 * The return value is ignored.
+	 */
+	validateArgs?: (args: HandlerArgs<NamedArgs>) => void | Promise<void>;
+
+	/**
+	 * The implementation of the command which is given camelCase'd args
+	 * and a ctx object of convenience properties
+	 */
+	handler: (
+		args: HandlerArgs<NamedArgs>,
+		ctx: HandlerContext
+	) => void | Promise<void>;
+};
+
+export const COMMAND_DEFINITIONS: Array<
+	CommandDefinition | NamespaceDefinition | AliasDefinition
+> = [];
```

Comment:
Yeah I did and it seemed like naming fatigue and lots of import/export for not much gain. Open to changing in the future but if they're only used in one place (`main()`) then this pattern makes sense.

Reply:
I was wondering this too. What if they're used in multiple places, like in separate isolated packages?

To avoid any potential issues with shared state, and to make it more portable, we could create an index file here and export functions with closed state: 

```
export const { defineCommand, defineNamespace, defineAlias } = createCommandDefineFns(); // terrible name...
```

This could then be created in each package where we want to isolate this behaviour, or could be used in the main wrangler package just as we're doing here in `kv`.